File: varg-compiler/benchmarks/benchmark.ts
// Equivalent benchmark to benchmark_realistic.varg in TypeScript

function fibonacci(n: number): number {
    let fa = 0, fb = 1;
    for (let i = 0; i < n; i++) {
        const temp = fb;
        fb = fa + fb;
        fa = temp;
    }
    return fa;
}

function sumRange(startVal: number, endVal: number): number {
    let total = 0;
    let idx = startVal;
    while (idx < endVal) {
        total += idx;
        idx += 1;
    }
    return total;
}

class Benchmark {
    counter = 0;

    runOnce(): void {
        // 1. Fibonacci(35)
        const fib = fibonacci(35);
        this.counter += fib;

        // 2. String building (10000 concats)
        let result = "";
        for (let i = 0; i < 10000; i++) {
            result += "x";
        }

        // 3. Array fill + sum (10000 elements)
        const numbers = [0];
        for (let i = 1; i < 10000; i++) {
            numbers.push(i);
        }
        let sum = 0;
        for (let i = 0; i < 10000; i++) {
            sum += numbers[i];
        }

        // 4. Nested loop (matrix-like 200x200)
        let matrixSum = 0;
        for (let row = 0; row < 200; row++) {
            for (let col = 0; col < 200; col++) {
                matrixSum += row * col;
            }
        }

        // 5. Sum range
        const rangeSum = sumRange(0, 1000);
        this.counter += rangeSum;
    }

    run(): void {
        console.log("=== TypeScript/Bun Benchmark (100 iterations) ===");
        for (let i = 0; i < 100; i++) {
            this.runOnce();
        }
        console.log(this.counter);
        console.log("=== Done ===");
    }
}

new Benchmark().run();

File: varg-compiler/benchmarks/benchmark.py
"""Equivalent benchmark to benchmark_realistic.varg in Python"""

def fibonacci(n: int) -> int:
    fa, fb = 0, 1
    for _ in range(n):
        fa, fb = fb, fa + fb
    return fa

def sum_range(start_val: int, end_val: int) -> int:
    total = 0
    idx = start_val
    while idx < end_val:
        total += idx
        idx += 1
    return total

class Benchmark:
    def __init__(self):
        self.counter = 0

    def run_once(self):
        # 1. Fibonacci(35)
        fib = fibonacci(35)
        self.counter += fib

        # 2. String building (10000 concats)
        result = ""
        for _ in range(10000):
            result += "x"

        # 3. Array fill + sum (10000 elements)
        numbers = [0]
        for i in range(1, 10000):
            numbers.append(i)
        total = sum(numbers)

        # 4. Nested loop (matrix-like 200x200)
        matrix_sum = 0
        for row in range(200):
            for col in range(200):
                matrix_sum += row * col

        # 5. Sum range
        range_sum = sum_range(0, 1000)
        self.counter += range_sum

    def run(self):
        print("=== Python Benchmark (100 iterations) ===")
        for _ in range(100):
            self.run_once()
        print(self.counter)
        print("=== Done ===")

if __name__ == "__main__":
    Benchmark().run()
